File: test/invoke_bot/perform_action/test_diagram_action_bar.js
/**
 * Panel tests: diagram action bar in workspace section view.
 * Tests rendering, button layout, workspace section structure.
 * Routes through CLI. Uses generic fixtures (EpicA, SubEpicB).
 */

// Mock vscode before requiring any modules
const Module = require('module');
const originalRequire = Module.prototype.require;
Module.prototype.require = function (...args) {
    if (args[0] === 'vscode') {
        return require('../../helpers/mock_vscode');
    }
    return originalRequire.apply(this, args);
};

const { test, after, before } = require('node:test');
const assert = require('assert');
const path = require('path');
const os = require('os');
const fs = require('fs');
const PanelView = require('../../../src/panel/panel_view');
const WorkspaceSectionView = require('../../../src/panel/workspace_section_view');

const repoRoot = path.join(__dirname, '../../..');
const productionBotPath = path.join(repoRoot, 'bots', 'story_bot');
const tempWorkspaceDir = fs.mkdtempSync(path.join(os.tmpdir(), 'agile-bots-diagram-action-bar-test-'));

function createMinimalStoryGraph() {
    return {
        epics: [
            {
                name: 'EpicA',
                sequential_order: 1,
                sub_epics: [
                    { name: 'SubEpicB', sequential_order: 1, sub_epics: [], story_groups: [] }
                ]
            }
        ]
    };
}

function setupTestWorkspace() {
    fs.mkdirSync(path.join(tempWorkspaceDir, 'docs', 'story'), { recursive: true });
    const storyGraphPath = path.join(tempWorkspaceDir, 'docs', 'story', 'story-graph.json');
    fs.writeFileSync(storyGraphPath, JSON.stringify(createMinimalStoryGraph(), null, 2), 'utf8');
    process.env.WORKING_AREA = tempWorkspaceDir;
    process.env.AGILE_BOTS_REPO_ROOT = repoRoot;
    const { verifyTestWorkspace } = require('../../helpers/prevent_production_writes');
    verifyTestWorkspace();
}

before(() => {
    setupTestWorkspace();
});

const cli = new PanelView(productionBotPath);

after(() => {
    cli.cleanup();
    if (fs.existsSync(tempWorkspaceDir)) {
        fs.rmSync(tempWorkspaceDir, { recursive: true, force: true });
    }
    delete process.env.AGILE_BOTS_REPO_ROOT;
    delete process.env.WORKING_AREA;
});

test('TestDiagramActionBarPanel', { concurrency: false }, async (t) => {
    await t.test('workspace_section_shows_diagram_buttons', async () => {
        await cli.execute('scope "EpicA"');
        const botData = await cli.execute('status');
        const view = new WorkspaceSectionView(cli);
        view.parentView = { botData };
        const html = await view.render();

        assert.ok(html.includes('ws-diagram-buttons'), 'Workspace section should have diagram buttons container');
        assert.ok(html.includes('alt="Render Diagram"'), 'Should show Render Diagram icon button');
        assert.ok(html.includes('alt="Save Layout"'), 'Should show Save Layout icon button');
        assert.ok(html.includes('alt="Clear Layout"'), 'Should show Clear Layout icon button');
        assert.ok(html.includes('alt="Generate Report"'), 'Should show Generate Report icon button');
        assert.ok(html.includes('alt="Update Graph"'), 'Should show Update Graph icon button');
        assert.ok(html.includes('renderDiagram'), 'Render button should post renderDiagram command');
        assert.ok(html.includes('saveDiagramLayout'), 'Save button should post saveDiagramLayout command');
        assert.ok(html.includes('clearDiagramLayout'), 'Clear button should post clearDiagramLayout command');
        assert.ok(html.includes('generateDiagramReport'), 'Generate Report button should post generateDiagramReport command');
        assert.ok(html.includes('updateFromDiagram'), 'Update button should post updateFromDiagram command');
    });

    await t.test('workspace_section_has_diagrams_and_build_subsections', async () => {
        await cli.execute('scope "EpicA"');
        const botData = await cli.execute('status');
        const view = new WorkspaceSectionView(cli);
        view.parentView = { botData };
        const html = await view.render();

        assert.ok(html.includes('Diagrams'), 'Workspace section should have Diagrams subsection');
        assert.ok(html.includes('Build'), 'Workspace section should have Build subsection');
        assert.ok(html.includes('Workspace'), 'Should have Workspace section header');
    });

    await t.test('workspace_section_has_open_graph_and_open_all_buttons', async () => {
        await cli.execute('scope "EpicA"');
        const botData = await cli.execute('status');
        const view = new WorkspaceSectionView(cli);
        view.parentView = { botData };
        const html = await view.render();

        assert.ok(html.includes('ws-btn-open-graph'), 'Build subsection should have Open Graph button');
        assert.ok(html.includes('ws-btn-open-all'), 'Build subsection should have Open All button');
        assert.ok(html.includes('alt="Graph"'), 'Open Graph button should have Graph alt text');
        assert.ok(html.includes('alt="All"'), 'Open All button should have All alt text');
    });

    await t.test('diagram_buttons_use_scope', async () => {
        await cli.execute('scope "EpicA"');
        const botData = await cli.execute('status');
        const view = new WorkspaceSectionView(cli);
        view.parentView = { botData };
        const html = await view.render();
        assert.ok(html.includes('window.diagramScope'), 'Diagram buttons should use window.diagramScope for scope');
    });
});

File: test/invoke_bot/perform_action/test_diagram_action_bar.py
"""
Test Diagram Action Bar

Domain tests: Bot API for diagram actions (get_diagram_action_bar_buttons, render, save, update).
CLI tests: render_diagram command routes correctly.
Panel tests: in test_diagram_action_bar.js (rendering, button layout).

Uses generic fixture names (EpicA, SubEpicB). No story-specific identifiers.
"""
import json
import pytest
from pathlib import Path
from scope.scope import ScopeType

from helpers.bot_test_helper import BotTestHelper
from helpers import TTYBotTestHelper, PipeBotTestHelper, JsonBotTestHelper


def _given_story_map_with_epic_and_subepic(tmp_path, epic_name="EpicA", subepic_name="SubEpicB"):
    """Given StoryMap has epic and sub-epic (generic names for domain/CLI tests)."""
    helper = BotTestHelper(tmp_path)
    helper.story.create_story_graph_with_child("Epic", epic_name, subepic_name)
    return helper


class TestDiagramActionBarDomain:
    """
    Domain tests: Bot returns diagram action bar buttons, render/save/update for scoped node.
    No CLI. Uses generic fixture names (EpicA, SubEpicB).
    """

    def test_bot_returns_diagram_action_bar_buttons_for_story_node(self, tmp_path):
        helper = _given_story_map_with_epic_and_subepic(tmp_path)
        node = helper.bot.story_map.find_node("SubEpicB")
        assert node is not None
        buttons = helper.bot.get_diagram_action_bar_buttons(node)
        assert "Render diagram" in buttons or "render" in str(buttons).lower()

    def test_bot_render_diagram_for_scope_returns_status(self, tmp_path):
        helper = _given_story_map_with_epic_and_subepic(tmp_path)
        helper.bot._scope.filter(ScopeType.STORY, ["SubEpicB"])
        helper.bot._scope.save()
        result = helper.bot.render_diagram_for_scope()
        assert "status" in result or "diagram" in str(result).lower()

    def test_bot_save_layout_to_drawio_returns_status_or_path(self, tmp_path):
        helper = _given_story_map_with_epic_and_subepic(tmp_path)
        result = helper.bot.save_layout_to_drawio()
        assert "status" in result or "path" in str(result).lower()

    def test_bot_update_graph_from_diagram_returns_status_or_report(self, tmp_path):
        helper = _given_story_map_with_epic_and_subepic(tmp_path)
        result = helper.bot.update_graph_from_diagram()
        assert "status" in result or "report" in str(result).lower()

    def test_bot_render_diagram_for_empty_scope_returns_status(self, tmp_path):
        helper = _given_story_map_with_epic_and_subepic(tmp_path)
        helper.bot._scope.filter(ScopeType.STORY, ["SubEpicB"])
        helper.bot._scope.save()
        result = helper.bot.render_diagram_for_scope()
        assert "status" in result
        assert result.get("status") in ("success", "complete") or "diagram" in str(result).lower()

    def test_bot_render_diagram_fails_when_scope_node_not_in_story_map(self, tmp_path):
        helper = BotTestHelper(tmp_path)
        helper.story.create_story_graph_with_child("Epic", "EpicA", "OtherNode")
        helper.bot._scope.filter(ScopeType.STORY, ["MissingNode"])
        helper.bot._scope.save()
        result = helper.bot.render_diagram_for_scope()
        assert result.get("status") == "error" or "error" in str(result).lower()


class TestDiagramActionBarCLI:
    """
    CLI tests: render_diagram command routes correctly.
    Uses generic scope path (EpicA, SubEpicB).
    """

    @pytest.mark.parametrize("helper_class", [TTYBotTestHelper, PipeBotTestHelper, JsonBotTestHelper])
    def test_cli_render_diagram_for_scope_reports_completion(self, tmp_path, helper_class):
        helper = helper_class(tmp_path)
        helper.domain.story.create_story_graph_with_child("Epic", "EpicA", "SubEpicB")
        cli_response = helper.cli_session.execute_command(
            'story_graph."EpicA"."SubEpicB".render_diagram'
        )
        output = cli_response.output
        if output.strip().startswith("{"):
            data = json.loads(output)
            assert data.get("status") in ("success", "complete")
        else:
            assert "success" in output.lower() or "complete" in output.lower()
